### ISSUE TITLE
Propagate Shortcut and KeyPress events to parent Canvas when a Widget is focused and it's not handling the Key/Shortcut

### ISSUE THREAD

**Issue by matwachich** [fyne-io/fyne#3038]
### Is your feature request related to a problem? Please describe:
Currently, when registering a Shortcut or a OnTypedKey on the Canvas (global), it will not be "fired" when a Widget is actualy focused, eventhough the Widget does nothing with the Shortcut or KeyPress.

### Is it possible to construct a solution with the existing API?
Nop

### Describe the solution you'd like to see:
The Canvas should have a chance to handle a global Shortcut eventhough some Widget is focused.

**Comment by Jacalz** [comment]
I think the idea here is good but I don’t think it would be possible to see if a widget has handled the key or not. The `KeyEvent` is just given as an input into the function and if a widget cares about it or not will be very hard (if not even impossible) to check.

I don’t know what that would be different for shortcuts because I can’t remember exactly how it works.

**Comment by matwachich** [comment]
This is a prerequisit to accelerators I think. So it should be worth some
code refactoring.
Why not adding a return value to callback functions, like a boolean that if
false, il will say that the key event should be passed to the Canvas

**Comment by matwachich** [comment]
Idea: a globally registered shortcut should never arrive at a widget, but it should be processed by the canvas. Like this, we are sure to receive the global shortcut where it should be processed, even if some widget is focused.

This comes to the cost that widgets will not receive some shortcuts, but I think this is logic since that shortcut has been registered as global.

**Comment by dweymouth** [comment]
> This comes to the cost that widgets will not receive some shortcuts

This seems unacceptable and would require some major workarounds. E.g. in my app, I have a canvas-level Ctrl-A shortcut that is used to select all selectable items in the current view. But I also have an Entry widget, and if that Entry has focus, Ctrl-A is expected to select all text in the entry, not to select everything selectable in the UI (as it does when entry isn't focused).

**Comment by Jacalz** [comment]
I personally think the easiest and best solution to solving this is to do a breaking change for v3.0.0 where the `Focusable` interface (or implementing a new interface perhaps, to avoid needing a breaking change?) returns a bool to specify if the shortcut was handled or not. It seems like a sensible way for a widget to say that it has caught a shortcut which can let the driver propagate the unused ones upwards and so on.

**Comment by matwachich** [comment]
Suggestion on the Shortcutable interface

```go
type Shortcutable interface {
    TypedShortcut(fyne.Shortcut) (handled bool)
}
```

We return true if shortcut has been handled by widget and should NOT be passed to the canvas, and the default false behavior remains that shortcut should propagate to the canvas.

**Comment by andydotxyz** [comment]
We can't just change the return parameter of an existing interface sorry, would have to find a way that added instead of replaced - maybe transitioning over time and removing in v3, or maybe not.

**Comment by Jacalz** [comment]
Indeed. We could add a second interface (with what name?) with the new method replacing the old one but otherwise like the old one. If that is implemented and it returns false, we can propagate the shortcut upwards. We just have to check the new interface before checking the existing one but all prior code will still work.

**Comment by chicogamedev** [comment]
Hi,

Very interesting conversation here.

I'm actually in the same issue. Actually I don't want to manage all the shortcuts in a custom entry type.

For example I have a Ctrl+N Shortcut. The Entry do nothing with it but it just eat it and do not propagate it. So I've created a CustomEntry type to manage the OnTypedShortcut on it. The problem is that there is no way to just say that if it's a custom shortcut, just call it on the canvas. The Shortcut handler of the canvas is not exported.

So that's a pretty bad design flaw in my opinion. I'm forced to find a way to avoid too much code duplication to call the same function already linked to the shortcuts on my canvas.

Maybe there could be a simpler approach to this issue by giving a way to manually invoke a shortcut on the window canvas.

What do you think ?


Thanks

**Comment by chicogamedev** [comment]
Yes I use Canvas.Addshortcut but these shortcuts are ignored when an entry is focused. That’s what my message is about.

it’s getting messy to tell my custom entry what todo with each of my custom shortcuts.
